File: solutions/nim/dnd-character/1/dnd_character.nim
import std/[math, random]

type
  Character* = object
    strength*: int
    dexterity*: int
    constitution*: int
    intelligence*: int
    wisdom*: int
    charisma*: int
    hitpoints*: int

randomize()
proc d6: int = rand(1..6)

proc ability*: int =
  var rolls = [d6(), d6(), d6(), d6()]
  sum(rolls) - min(rolls)

proc modifier*(n: int): int =
  floorDiv(n - 10, 2)

proc initCharacter*: Character =
  let con = ability()

  Character(
    strength: ability(),
    dexterity: ability(),
    constitution: con,
    intelligence: ability(),
    wisdom: ability(),
    charisma: ability(),
    hitpoints: 10 + modifier(con)
  )

File: solutions/csharp/dnd-character/1/DndCharacter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DndCharacter
{
    public int Strength { get; private set; }
    public int Dexterity { get; private set; }
    public int Constitution { get; private set; }
    public int Intelligence { get; private set; }
    public int Wisdom { get; private set; }
    public int Charisma { get; private set; }
    public int Hitpoints { get; private set; }

    public static int Modifier(int score)
    {
        return (int)Math.Floor((score-10)/2.0);
    }

    private static Random _rand = new Random();

    public static int Ability() 
    {
        return Enumerable.Range(1, 4)
            .Select(_ => _rand.Next(1,7))
            .OrderByDescending(x => x)
            .Take(3)
            .Sum();
    }

    public static DndCharacter Generate()
    {
        var constitution = Ability();
        var constitutionMod = Modifier(constitution);
        
        return new DndCharacter() {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = constitution,
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability(),
            Hitpoints = 10 + constitutionMod
        };
    }
}
